### ISSUE TITLE
Schema-less while avoiding incoherent structure

### ISSUE THREAD

**Issue by SamyPesse** [ianstormtaylor/slate#206]
I'm working on a large rich text editor (similar to Google Docs), including lists, code blocks, blockquotes, tables, etc.

We were previously using Prosemirror, which is build around a [schema](https://prosemirror.net/guide/schema.html).

First of all **I agree that the core of Slate should be Schema-less**.

But with a large rich text editor, we have to ensure some kind of structure:
- Lists can only contain `list_item`
- Tables can only contain `table_row`
- Rows can only contain `table_cell`
- Cells can only contain text

The transformations on the document should avoid breaking this structure, but it can easily add complexity to handle edge cases.

I'm looking for proposition on how to solve this issue with the best and cleaner solution.

------

### Proposition 1: `slate-schema` plugin

Basically, a plugin `slate-schema` can ensure a structure on `onChange`:

**Plus**: When `onChange` is called, the document always has the right structure
**Minus:** Performances?

```js
import Schema from 'slate-schema'

const schema = Schema({
    "list_ul": {
          only: ["list_item"],
          min: 1
    },
    "list_ol": {
          only: ["list_item"],
          min: 1
    },
    "table": {
          only: ["table_row"],
          min: 1
    },
    "table_row": {
          only: ["table_cell"],
          min: 1
    }
})

const plugins = [
    schema
]
```

**Comment by ianstormtaylor** [comment]
Having a `slate-schema` plugin is actually an awesome idea, I hadn't thought of that. I agree that this is the bigger downside to not having a schema in core.

For performance, you could use `onDocumentChange` for a bit better. And there might be something else we discover that can speed things up even more.

---

Another thing to consider: right now without a schema, technically if someone pastes from another Slate instance that has a different schema into another, whatever the blocks/inlines/data/marks that were defined in that first Slate will be added to the second. This might also be a good place for the schema plugin to listen `onPaste` or `onDrop` to ensure that when `data.type == 'fragment'`, the `data.fragment` is transformed to ensure that it conforms.

(For that last case, there could be another solution where `Document` would have `data`, which would be inherited by the `fragments` created from it, so you could then decide based on say `data.isMyEditor` whether to accept the `fragment` or fallback to `text`.)

**Comment by SamyPesse** [comment]
I've publish a first PoC for `slate-schema`: https://github.com/SamyPesse/slate-schema

It'd be great to have your feedback on the schema format, etc.

Also I'm not sure how to return a new state from `onDocumentChange`, I've seen [examples/images/index.js#L125](https://github.com/ianstormtaylor/slate/blob/d42c463353b164ab7d3a4a583010c70295b30c1e/examples/images/index.js#L125), but I don't how to call the `onChange` from the plugin.

I'll run some real tests in our application tomorrow to see what is missing, performances, etc

**Comment by ianstormtaylor** [comment]
Ah cool! For `onDocumentChange`, that's mostly a helper for saving/syncing data on changes. Right now there's no plugin-level version of it, although maybe there should be. I'd be down to accept a PR that added that!

**Comment by SamyPesse** [comment]
I'll try to start a PR for this.

Do you agree that `onDocumentChange` should be able to return a new document or `undefined`? or do you have something else in mind?

**Comment by ianstormtaylor** [comment]
Actually, that's a good point, it's not really clear. I think for now it makes more sense to keep it as a plugin-level concern. Basically you'd just need to keep track of `previousDocument` in the closure, and check to see if the document is different. Something like:

```js
function MyPlugin(options) {
  let previousDocument 

  return {
    onChange(state) {
      if (state.document == previousDocument) return

      previousDocument = state.document
      ...
    }
  }
}
```
